File: sdk-java-blocking/src/test/java/dev/restate/sdk/blocking/InvocationIdTest.java
package dev.restate.sdk.blocking;

import static dev.restate.sdk.core.impl.ProtoUtils.greetingResponse;

import dev.restate.sdk.core.InvocationId;
import dev.restate.sdk.core.impl.InvocationIdTestSuite;
import dev.restate.sdk.core.impl.testservices.GreeterGrpc;
import dev.restate.sdk.core.impl.testservices.GreetingRequest;
import dev.restate.sdk.core.impl.testservices.GreetingResponse;
import io.grpc.BindableService;
import io.grpc.stub.StreamObserver;

public class InvocationIdTest extends InvocationIdTestSuite {

  private static class ReturnInvocationId extends GreeterGrpc.GreeterImplBase
      implements RestateBlockingService {

    @Override
    public void greet(GreetingRequest request, StreamObserver<GreetingResponse> responseObserver) {
      responseObserver.onNext(greetingResponse(InvocationId.current().toString()));
      responseObserver.onCompleted();
    }
  }

  @Override
  protected BindableService returnInvocationId() {
    return new ReturnInvocationId();
  }
}

File: sdk-kotlin/src/test/kotlin/dev/restate/sdk/kotlin/InvocationIdTest.kt
package dev.restate.sdk.kotlin

import dev.restate.sdk.core.InvocationId
import dev.restate.sdk.core.impl.InvocationIdTestSuite
import dev.restate.sdk.core.impl.testservices.GreeterGrpcKt
import dev.restate.sdk.core.impl.testservices.GreetingRequest
import dev.restate.sdk.core.impl.testservices.GreetingResponse
import dev.restate.sdk.core.impl.testservices.greetingResponse
import io.grpc.BindableService
import kotlinx.coroutines.Dispatchers

class InvocationIdTest : InvocationIdTestSuite() {
  private class ReturnInvocationId :
      GreeterGrpcKt.GreeterCoroutineImplBase(Dispatchers.Unconfined), RestateCoroutineService {
    override suspend fun greet(request: GreetingRequest): GreetingResponse {
      return greetingResponse { message = InvocationId.current().toString() }
    }
  }

  override fun returnInvocationId(): BindableService {
    return ReturnInvocationId()
  }
}
